### PR TITLE
[SPARK-33020][PYTHON] Add nth_value as a PySpark function

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

`nth_value` was added at SPARK-27951. This PR adds the corresponding PySpark API.

### Why are the changes needed?

To support the consistent APIs

### Does this PR introduce _any_ user-facing change?

Yes, it introduces a new PySpark function API.

### How was this patch tested?

Unittest was added.

### CODE REVIEW

File: python/pyspark/sql/functions.pyi [hunk 1]
```diff
@@ -85,6 +85,9 @@ def lag(
 def lead(
     col: ColumnOrName, offset: int = ..., default: Optional[Any] = ...
 ) -> Column: ...
+def nth_value(
+    col: ColumnOrName, offset: int, ignoreNulls: Optional[bool] = ...
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.pyi:89]: I was about to suggest to make it accept `Column` in the original PR but .. let's leave it matched for now - I think Window once might have to be treated differently given that it has different SQL syntax.

File: python/pyspark/sql/functions.pyi [hunk 2]
```diff
@@ -85,6 +85,9 @@ def lag(
 def lead(
     col: ColumnOrName, offset: int = ..., default: Optional[Any] = ...
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.pyi:86]: I'll take a look separately for these primitive types vs `Column` issue.

**Comment by HyukjinKwon** [comment]
cc @zero323 would you mind taking a look when you find some time?

File: python/pyspark/sql/functions.py [hunk 3]
```diff
@@ -934,6 +934,25 @@ def lead(col, offset=1, default=None):
...
+    true. If all values are null, then null is returned.
+
+    This is equivalent to the nth_value function in SQL.
+
+    :param col: name of column or expression
+    :param offset: number of row to use as the value
+    :param ignoreNulls: default value
+    """
+    sc = SparkContext._active_spark_context
+    return Column(sc._jvm.functions.nth_value(_to_java_column(col), offset, ignoreNulls))
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:953]: Ideally we should also do the type check. But let me just match it with other window functions for now, and do it separately together with other window functions.

**Comment by HyukjinKwon** [comment]
cc @beliefer and @cloud-fan FYI

File: python/pyspark/sql/functions.py [hunk 4]
```diff
@@ -934,6 +934,25 @@ def lead(col, offset=1, default=None):
...
+    (counting from 1), and `null` if the size of window frame is less than `offset` rows.
+
+    It will return the `offset`\\th non-null value it sees when `ignoreNulls` is set to
+    true. If all values are null, then null is returned.
+
+    This is equivalent to the nth_value function in SQL.
+
+    :param col: name of column or expression
+    :param offset: number of row to use as the value
+    :param ignoreNulls: default value
```
Comment (beliefer) [inline python/pyspark/sql/functions.py:950]: an optional specification that indicates the NthValue should skip null values in the determination of which row to use.

File: python/pyspark/sql/functions.py [hunk 5]
```diff
@@ -934,6 +934,25 @@ def lead(col, offset=1, default=None):
     return Column(sc._jvm.functions.lead(_to_java_column(col), offset, default))
 
 
+@since(3.1)
+def nth_value(col, offset, ignoreNulls=False):
+    """
+    Window function: returns the value that is the `offset`\\th row of the window frame
```
Comment (beliefer) [inline python/pyspark/sql/functions.py:940]: `\\th`?

**Comment by beliefer** [comment]
@HyukjinKwon Thank you! I learned more about PySpark.

File: python/pyspark/sql/functions.py (hunk 5 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/functions.py:940]: Yeah, it should have `\\` to distinguish `th` from the previous backquote.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @HyukjinKwon .

**Comment by dongjoon-hyun** [comment]
Merged to master.

**Comment by dongjoon-hyun** [comment]
Merged to master.

**Comment by HyukjinKwon** [comment]
Thank you @dongjoon-hyun.

**Comment by zero323** [comment]
> cc @zero323 would you mind taking a look when you find some time?

Looks good. I am going to add this to R as well, unless there someone already works on that.

**Comment by HyukjinKwon** [comment]
Thanks @zero323 please go ahead!

**Pull request closed without merging**